File: tests/python/test-gegl-color.py
#!/usr/bin/env python
""" This file is part of GEGL
 *
 * GEGL is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * GEGL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with GEGL; if not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2013 Daniel Sabo <DanielSabo@gmail.com>
"""

import unittest

from gi.repository import Gegl

class TestGeglColor(unittest.TestCase):
    def assertAlmostEqualComps(self, a, b):
        self.assertEqual(len(a), len(b))
        for acomp, bcomp in zip(a, b):
            self.assertAlmostEqual(acomp, bcomp)

    def test_new_color(self):
        Gegl.Color.new("rgba(0.6, 0.6, 0.6, 1.0)")

    def test_new_color_string(self):
        Gegl.Color(string="rgba(0.6, 0.6, 0.6, 1.0)")

    def test_color_set_rgba(self):
        c = Gegl.Color.new("rgba(1.0, 1.0, 1.0, 1.0)")
        values = c.get_rgba()
        self.assertAlmostEqual(values[0], 1.0)
        self.assertAlmostEqual(values[1], 1.0)
        self.assertAlmostEqual(values[2], 1.0)
        self.assertAlmostEqual(values[3], 1.0)
        c.set_rgba(0.3, 0.6, 0.9, 1.0)
        values = c.get_rgba()
        self.assertAlmostEqual(values[0], 0.3)
        self.assertAlmostEqual(values[1], 0.6)
        self.assertAlmostEqual(values[2], 0.9)
        self.assertAlmostEqual(values[3], 1.0)

    def test_color_get_components(self):
        c = Gegl.Color()
        c.set_components(Gegl.format("RGB float"), [1.0, 0.0, 0.0])

        values = c.get_components(Gegl.format("RGB float"))
        self.assertAlmostEqualComps(values, [1.0, 0.0, 0.0])

        values = c.get_components(Gegl.format("RGBA double"))
        self.assertAlmostEqualComps(values, [1.0, 0.0, 0.0, 1.0])

        values = c.get_components(Gegl.format("RGBA float"))
        self.assertAlmostEqualComps(values, [1.0, 0.0, 0.0, 1.0])

        values = c.get_components(Gegl.format("RGBA u32"))
        self.assertEqual(values, [float(0xFFFFFFFF), 0.0, 0.0, float(0xFFFFFFFF)])

        values = c.get_components(Gegl.format("RGBA u16"))
        self.assertEqual(values, [float(0xFFFF), 0.0, 0.0, float(0xFFFF)])

        values = c.get_components(Gegl.format("RGBA u8"))
        self.assertEqual(values, [float(0xFF), 0.0, 0.0, float(0xFF)])

        c.set_components(Gegl.format("R'G'B' u8"), [128, 0, 128])

        values = c.get_components(Gegl.format("R'G'B'A u8"))
        self.assertEqual(values, [float(128), 0.0, float(128), float(255)])

        c.set_components(Gegl.format("YA double"), [0.5, 0.5])

        values = c.get_components(Gegl.format("RGBA double"))
        self.assertAlmostEqualComps(values, [0.5, 0.5, 0.5, 0.5])

        values = c.get_components(Gegl.format("RaGaBaA double"))
        self.assertAlmostEqualComps(values, [0.25, 0.25, 0.25, 0.5])

if __name__ == '__main__':
    Gegl.init(None);
    unittest.main()
    Gegl.exit()

File: tests/simple/test-gegl-color.c
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2014 Jon Nordby <jononor@gmail.com>
 */


#include <glib.h>
#include <gegl.h>
#include <math.h>
#include <stdlib.h>

/* Common */

#define COLOR_FMT "%.3f"
static const gdouble color_diff_threshold = 0.001;

static void
assert_compare_rgba(const gdouble actual[4], const gdouble expect[4])
{
    gdouble diff[4];

    gboolean colors_equal = TRUE;
    for (int i=0; i<4; i++) {
        diff[i] = fabs(actual[0] - expect[0]);
        if (diff[i] > color_diff_threshold) {
            colors_equal = FALSE;
        }
    }

    if (!colors_equal) {
        g_error("GeglColor("COLOR_FMT","COLOR_FMT","COLOR_FMT","COLOR_FMT") != "
                "GeglColor("COLOR_FMT","COLOR_FMT","COLOR_FMT","COLOR_FMT")",
                actual[0], actual[1], actual[2], actual[3],
                expect[0], expect[1], expect[2], expect[3]);
    }
    g_assert(colors_equal);
}

/* Test GeglColor set_rgba/get_rgba roundtrip */
#define TP_ROUNDTRIP "/GeglColor/rgba_set_get/"

typedef struct _TestRgbaSetGetData {
    gchar *casename;
    gdouble rgba[4];
} TestRgbaSetGetData;

static TestRgbaSetGetData roundtrip_cases[] = {
    { TP_ROUNDTRIP"all_components_zero", { 0., 0., 0., 0. } },
    { TP_ROUNDTRIP"all_components_one", { 1., 1., 1., 1. } },
    { TP_ROUNDTRIP"mixed_components", { 0.3, 0.1, 0.5, 0.7 } }
};

static void
test_rgba_setget_roundtrip(const TestRgbaSetGetData *d)
{
    gdouble out[4] = { -1, -1, -1, -1 };

    GeglColor *color = gegl_color_new(NULL);
    gegl_color_set_rgba(color, d->rgba[0], d->rgba[1], d->rgba[2], d->rgba[3]);
    gegl_color_get_rgba(color, &out[0], &out[1], &out[2], &out[3]);
    assert_compare_rgba(out, d->rgba);
    g_object_unref(color);
}

/* Test GeglColor parsing of color strings */
#define TP_PARSE "/GeglColor/parse/"

typedef struct _TestParseString {
    gchar *casename;
    gchar *input;
    gdouble expected[4];
} TestParseString;

static TestParseString parse_cases[] = {
    /* cannot be tested because gegl_color_set_from_string causes g_warning 
    { TP_PARSE"invalid", "s22003",  { 0.f, 1.f, 1.f, 0.67f } }, */

    { TP_PARSE"named_white", "white",  { 1.f, 1.f, 1.f, 1.f } },
    { TP_PARSE"named_black", "black",  { 0.f, 0.f, 0.f, 1.f } },
    { TP_PARSE"named_olive", "olive",  { 0.21586f, 0.21586f, 0.f, 1.f } },

    { TP_PARSE"rgb_hex6_black", "#000000",  { 0.f, 0.f, 0.f, 1.f } },
    { TP_PARSE"rgb_hex6_white_lowercase", "#ffffff",  { 1.f, 1.f, 1.f, 1.f } },
    { TP_PARSE"rgb_hex6_white_uppercase", "#FFFFFF",  { 1.f, 1.f, 1.f, 1.f } },
    { TP_PARSE"rgb_hex8_olive_semitrans", "#808000aa",  { 0.21586f, 0.2195f, 0.f, 0.6666f } },
    { TP_PARSE"rgb_hex3_", "#333",  { 0.0331f, 0.0331f, 0.0331f, 1.f } },
    { TP_PARSE"rgb_hex4_", "#ccce",  { 0.6038f, 0.6038f, 0.6038f, 0.8549f } },

    // Note: not CSS compatible!
    { TP_PARSE"rgb_integer_black", "rgb(0,0,0)",  { 0.f, 0.f, 0.f, 1.f } },
    { TP_PARSE"rgb_float_white", "rgb(1.0, 1.0, 1.0)",  { 1.f, 1.f, 1.f, 1.f } },
    { TP_PARSE"rgba_middle_semitrans", "rgba(0.660, 0.330, 0.200, 0.4)",  { 0.660f, 0.330f, 0.200f, 0.4f } }
};

static void
test_parse_string(const TestParseString *data)
{
    gdouble out[4] = { -1, -1, -1, -1 };

    GeglColor *color = gegl_color_new(data->input);
    gegl_color_get_rgba(color, &out[0], &out[1], &out[2], &out[3]);
    assert_compare_rgba(out, data->expected);
    g_object_unref(color);
}


int
main (int argc, char *argv[])
{
    int result = -1;
    gegl_init(&argc, &argv);

    g_test_init(&argc, &argv, NULL);

    // set/get roundtrip tests
    for (int i = 0; i < G_N_ELEMENTS(roundtrip_cases); i++) {
        const TestRgbaSetGetData *data = &roundtrip_cases[i];
        g_test_add_data_func(data->casename, data, (GTestDataFunc)test_rgba_setget_roundtrip);
    }

    // parsing tests
    for (int i = 0; i < G_N_ELEMENTS(parse_cases); i++) {
        const TestParseString *data = &parse_cases[i];
        g_test_add_data_func(data->casename, data, (GTestDataFunc)test_parse_string);
    }

    result = g_test_run();
    gegl_exit();
    return result;
}
